tests/integration/ef_tests/runner.py

```diff
@@ -0,0 +1,143 @@
+):
+    """
+    Initialize the Starknet state using an ef-test BlockchainTest format case 'pre' field.
+
+    See: https://ethereum-tests.readthedocs.io/en/latest/test_filler/blockchain_filler.html#pre for details.
+    """
+    for address, state in pre_state.items():
+        evm_address = int(address, 16)
+        starknet_address = get_starknet_address(evm_address)
+
+        balance = int_to_uint256(int(state["balance"], 16))
+        await fund_and_set_allowance(
+            eth, kakarot.contract_address, starknet_address, balance
+        )
+
+        if is_account_eoa(state):
+            await write_test_state_for_eoa(
+                externally_owned_account_class.class_hash,
+                account_proxy_class.class_hash,
+                kakarot,
+                starknet,
+                state,
+                evm_address,
+                starknet_address,
+            )
+        else:
+            contract = get_contract_account(starknet_address)
+            await write_test_state_for_contract(
+                contract_account_class.class_hash,
+                account_proxy_class.class_hash,
+                contract,
+                kakarot,
+                starknet,
+                state,
+                evm_address,
+                starknet_address,
+            )
+
+
+async def do_transaction(
+    blocks,
+    get_starknet_address,
+    kakarot: StarknetContract,
+    owner,
+    starknet: StarknetContract,
+):
+    """
+    Execute each transaction in each block using an ef-test BlockchainTest format's `blocks` field.
+
+    See: https://ethereum-tests.readthedocs.io/en/latest/test_filler/blockchain_filler.html#blocks for details.
+    """
+    for block in blocks:
+        transactions = block["transactions"]
+
+        for transaction in transactions:
+            starknet_address = get_starknet_address(int(transaction["sender"], 16))
+
+            await kakarot.eth_send_transaction(
+                to=int(transaction["to"], 16),
+                gas_limit=int(transaction["gasLimit"], 16),
+                gas_price=int(transaction["gasPrice"], 16),
+                value=int(transaction["value"], 16),
+                data=hex_string_to_bytes_array(transaction["data"]),
+            ).execute(caller_address=starknet_address)
+
+            # do we really have to do this manually? (yes)
+            await starknet.state.state.increment_nonce(starknet_address)
+
+
+async def assert_eoa_post_state(
+    address, starknet, starknet_address, expected_post_state
+):
+    actual_nonce = await starknet.state.state.get_nonce_at(starknet_address)
+    expected_nonce = int(expected_post_state["nonce"], 16)
+    assert (
+        actual_nonce == expected_nonce
+    ), f"EOA {address}: {expected_nonce=} is not {actual_nonce=}"
+    assert not expected_post_state[
+        "storage"
+    ], f"EOA {address} should have empty post state, but got {expected_post_state['storage']}"
+
+
+async def assert_contract_post_state(
```

Comment:
so the only reason why you split in two methods is to get the nonce? because otherwise I don't see the need for `assert_eoa_` and `assert_contract`